cmd/kops/delete_instance.go

```diff
@@ -0,0 +1,277 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/kops/cmd/kops/util"
+	kopsapi "k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/cloudinstances"
+	"k8s.io/kops/pkg/instancegroups"
+	"k8s.io/kops/pkg/validation"
+	"k8s.io/kops/upup/pkg/fi/cloudup"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+// DeleteInstanceOptions is the command Object for an instance deletion.
+type DeleteInstanceOptions struct {
+	CloudOnly bool
+
+	// The following two variables are when kops is validating a cluster
+	// between detatch and deletion.
+
+	// FailOnDrainError fail deletion if drain errors.
+	FailOnDrainError bool
+
+	// FailOnValidate fail the deletion when the cluster
+	// does not validate, after a validation period.
+	FailOnValidate bool
+
+	// PostDrainDelay is the duration of a pause after a drain operation
+	PostDrainDelay time.Duration
+
+	// ValidationTimeout is the timeout for validation to succeed after the drain and pause
+	ValidationTimeout time.Duration
+
+	// ValidateCount is the amount of time that a cluster needs to be validated between drain and deletion
+	ValidateCount int32
+
+	ClusterName string
+
+	InstanceID string
+
+	Detatch bool
+}
+
+func (o *DeleteInstanceOptions) InitDefaults() {
+	o.CloudOnly = false
+	o.FailOnDrainError = false
+	o.FailOnValidate = true
+
+	o.PostDrainDelay = 5 * time.Second
+	o.ValidationTimeout = 15 * time.Minute
+	o.ValidateCount = 2
+
+	o.Detatch = true
+}
+
+func NewCmdDeleteInstance(f *util.Factory, out io.Writer) *cobra.Command {
+	deleteIgLong = templates.LongDesc(i18n.T(`
+		Delete an instance. By default, it will detatch the instance from 
+		the instance group, drain it, then terminate it.`))
+
+	deleteIgExample = templates.Examples(i18n.T(`
+		# Delete an instance from the currently active cluster.
+		kops delete instance i-0a5ed581b862d3425
+
+		# Delete an instance from the currently active cluster without
+		validation or draining.
+		kops delete instance --cloudonly i-0a5ed581b862d3425
+		`))
+
+	deleteIgShort = i18n.T(`Delete instancegroup`)
```

Comment:
there are some copy/paste references to `instancegroup` including `Ig` in variable names.

Reply:
Should be cleaned up now.